modules/aggregation-layers/src/heatmap-layer/heatmap-layer.js

```diff
@@ -0,0 +1,484 @@
+  radiusPixels: {type: 'number', min: 1, max: 100, value: 30},
+  colorRange: defaultColorRange,
+  enhanceFactor: {type: 'number', min: 1, max: 100, value: 20}
+};
+
+export default class HeatmapLayer extends CompositeLayer {
+  initializeState() {
+    const {gl} = this.context;
+    if (!isWebGL2(gl)) {
+      log.error('HeatmapLayer is not supported on this browser, requires WebGL2')();
+    }
+    this._setupAttributes();
+    this._setupResources();
+  }
+
+  shouldUpdateState({changeFlags}) {
+    // Need to be updated when viewport changes
+    return changeFlags.somethingChanged;
+  }
+
+  updateState(opts) {
+    super.updateState(opts);
+    const {props, oldProps} = opts;
+    const changeFlags = this._getChangeFlags(opts);
+
+    if (changeFlags.viewportChanged) {
+      changeFlags.boundsChanged = this._updateBounds(changeFlags.viewportZoomChanged);
+    }
+
+    if (changeFlags.dataChanged) {
+      this._updateWeightmapAttributes();
+    }
+
+    if (changeFlags.dataChanged || changeFlags.boundsChanged || changeFlags.uniformsChanged) {
+      this._updateWeightmap();
+    }
+
+    if (props.colorRange !== oldProps.colorRange) {
+      this._updateColorTexture(opts);
+    }
+
+    // TODO: need to be fixed, seem to be rendring wrong part of the texture
+    if (changeFlags.viewportChanged) {
+      this._updateTextureRenderingBounds();
+    }
+
+    this.setState({zoom: opts.context.viewport.zoom});
+  }
+
+  renderLayers() {
+    const {
+      weightsTexture,
+      triPositionBuffer,
+      triTexCoordBuffer,
+      maxWeightsTexture,
+      colorTexture
+    } = this.state;
+    const {updateTriggers, enhanceFactor} = this.props;
+
+    return new TriangleLayer(
+      this.getSubLayerProps({
+        id: `${this.id}-triangle-layer`,
+        updateTriggers
+      }),
+      {
+        id: 'heatmap-triangle-layer',
+        data: {
+          attributes: {
+            positions: triPositionBuffer,
+            texCoords: triTexCoordBuffer
+          }
+        },
+        count: 6,
+        maxTexture: maxWeightsTexture,
+        colorTexture,
+        texture: weightsTexture,
+        opacityFactor: enhanceFactor
+      }
+    );
+  }
+
+  finalizeState() {
+    super.finalizeState();
+    const {
+      weightsTransform,
+      weightsTexture,
+      maxWeightsTexture,
+      triPositionBuffer,
+      triTexCoordBuffer
+    } = this.state;
+    weightsTransform.delete();
+    weightsTexture.delete();
+    maxWeightsTexture.delete();
+    triPositionBuffer.delete();
+    triTexCoordBuffer.delete();
+  }
+
+  // PRIVATE
+
+  // override Composite layer private method to create AttributeManager instance
+  _getAttributeManager() {
+    return new AttributeManager(this.context.gl, {
+      id: this.props.id,
+      stats: this.context.stats
+    });
+  }
+
+  _getChangeFlags(opts) {
+    const {oldProps, props} = opts;
+    const changeFlags = {};
+    if (this._isDataChanged(opts)) {
+      changeFlags.dataChanged = true;
```

Comment:
Try not to overwrite existing flags. This may have unexpected consequences.

Reply:
Only changing local object, `changeFlags`.